### DISCUSSION TITLE
Strongswan status freezes with HA configuration

### DISCUSSION THREAD

**Discussion by khramov86** [strongswan/strongswan#450] [Q&A] [answered]
Hi,

while testing HA configuration faced with problem, after 8+ hours that tunnel is up, in some moment of time service on passive host takes weight as on active (also tried implement active-active configuration and obtained same behaviour) and after that command `strongswan status` on secondary hosts don't work.
Strace of strongswan status in attachment with strongswan.log for primary and secondary.

[strongswan_primary.log](https://github.com/strongswan/strongswan/files/6660668/strongswan_primary.log)
[strongswan_secondary.log](https://github.com/strongswan/strongswan/files/6660669/strongswan_secondary.log)
[strongswan_status_strace.log](https://github.com/strongswan/strongswan/files/6660656/strongswan_status_strace.log)

Can you help me please?

**Comment by tobiasbrunner** [comment, marked as answer]
>  Strace of strongswan status in attachment

Not that helpful because not only is `ipsec` (or `strongswan` in your case) just a script, the only thing the called `stroke status` command does is try  to connect to a socket, so strace won't really tell you much.

> with strongswan.log for primary and secondary.

As I told you before, the HA plugin is not intended to be used on initiators (i.e. you can't use `auto=start` or `auto=route` or `dpd|closeaction=restart|hold`). According to the log, you have not changed that. It's possible that the issue you are seeing is related to that somehow. But I really have no idea what the specific problem could be.

**Reply by khramov86** [reply]
Hi, Tobias.

Thank you for response! Also tried with `auto=add` before, and get the same result. 
Generally, is this correct to use this parameters?
```
  auto=add
  dpdaction=clear
  dpddelay=35s
  dpdtimeout=300s
```

**Reply by tobiasbrunner** [reply]
I guess, but `dpdtimeout` is only used for IKEv1 and `dpddelay` might depend on your use case (e.g. how long you allow clients to roam around without network connectivity).

If this happens, does the daemon log anything at all? Or is it stuck completely? You could try increasing log levels (e.g. for _job_ or _cfg_ to see more about the config requests). And if it actually is stuck, try attaching GDB to the daemon and get backtraces for all threads (`thread apply all bt`).

**Reply by khramov86** [reply]
Hi, Tobias. When this happens I didn't see anything extraordinary in system log, also when I'm trying to execute `/usr/libexec/strongswan/stroke status` - command still stuck.
If this is correct, i've tryied to attach to PID of command that stuck `strongswan status &` with gdb and execute `thread apply all bt`

```
(gdb) thread apply all bt

Thread 1 (Thread 0x7f0f34979880 (LWP 25464)):
#0  0x00007f0f3422a7bd in recv () from /lib64/libc.so.6
#1  0x00007f0f3452fe8a in read_ () from /usr/lib64/strongswan/libstrongswan.so.0
#2  0x00005612a6602342 in send_stroke_msg ()
#3  0x00005612a66017c5 in main ()
(gdb)
```
Or I should get backtraces from charon/starter process?

**Reply by tobiasbrunner** [reply]
> Or I should get backtraces from charon/starter process?

Only `charon` is relevant. `starter` just starts that and parses the legacy `ipsec.conf` file and sends configs to `charon`. And as I explained above, `strongswan` is just a shell script (`status` calls the legacy `stroke` utility but, as you can see, that's just stuck reading from the socket the `stroke` plugin in `charon` has opened).

**Reply by khramov86** [reply]
Thank you, Tobias.
Backtraces from charon
[charon_thread_apply_all_bt.log](https://github.com/strongswan/strongswan/files/6676057/charon_thread_apply_all_bt.log)

**Reply by tobiasbrunner** [reply]
OK, I see what the problem is. It's a bug that was introduced with 5.9.2. See #456 for details.

**Reply by khramov86** [reply]
More than 24 hours HA cluster is working without problems.
Thank you, Tobias!

**Reply by tobiasbrunner** [reply]
OK, great. The fix is now in master.
